src/Microsoft.Identity.Client/TokenCache.cs

```diff
@@ -190,12 +162,12 @@ public bool HasStateChanged
 
         internal void OnAfterAccess(TokenCacheNotificationArgs args)
         {
-            AfterAccess?.Invoke(args);
+            _asyncAfterAccess?.Invoke(args).GetAwaiter().GetResult();
         }
 
         internal void OnBeforeAccess(TokenCacheNotificationArgs args)
         {
-            BeforeAccess?.Invoke(args);
+            _asyncBeforeAccess?.Invoke(args).GetAwaiter().GetResult();
```

Comment:
This will block the thread until the result is back. If we were to await instead, the CLR would release the thread and spawn another thread when the operation completes. So we have to await, which means OnAfterAccess becomes OnAfterAccessAsync. 
https://montemagno.com/c-sharp-developers-stop-calling-dot-result/

The argument for doing this work is to release the thread, which will lead to smaller contention on ASP.

The tricky bit is that you can't have 
```csharp
lock (obj)
{
    await OnAfterAccess(...);
}

because the lock would block re-entry.

Reply:
yeah, that's true.  if we're trying to get a semantically pretty API so that developers can call "await myFunctionAsync()" then i think what we have is right / close to right.  but we have to do this within the lock for synchronization reasons so you're correct that we can't do the lock (onafteraccess).  

Can we do a Task.Run() here and block on its completion so they get another thread but we block this thread on it to return?

Reply:
SemaphoreSlim can be used instead of a lock, it blocks access to the critical section. 

```csharp
await semaphoreSlim.WaitAsync();
try
{
    await Task.Delay(1000);
}
finally
{    
    semaphoreSlim.Release();
}
```

See https://blog.cdemi.io/async-waiting-inside-c-sharp-locks/